spark/src/main/scala/org/apache/spark/smile/SparkDataTypes.scala

```diff
@@ -0,0 +1,103 @@
+/*******************************************************************************
+ * Copyright (c) 2010-2020 Haifeng Li. All rights reserved.
+ *
+ * Smile is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version.
+ *
+ * Smile is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Smile.  If not, see <https://www.gnu.org/licenses/>.
+ ******************************************************************************/
+
+package org.apache.spark.smile
+
+import smile.data.`type`.{DataType, DataTypes, StructField, StructType}
+import org.apache.spark.sql.types._
+import org.apache.spark.ml.linalg.VectorUDT
+import org.apache.spark.mllib.linalg.{VectorUDT => OldVectorUDT}
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+
+object SparkDataTypes {
```

Comment:
I guess that classes in org.apache.spark.smile package don't use spark's private features. If so, let's move them to smile namespace.

Reply:
VectorUDT (which is the features DataType used by Spark Estimators) is private Spark as the UserDefineType.